locations/spiders/pick_n_pay.py

```diff
@@ -0,0 +1,49 @@
+import scrapy
+
+from locations.dict_parser import DictParser
+from locations.hours import DAYS, OpeningHours
+
+
+class PickNPaySpider(scrapy.Spider):
+    name = "pick_n_pay"
+    item_attributes = {"brand": "Pick n Pay", "brand_wikidata": "Q7190735"}
+    start_urls = ["https://api.pnp.co.za/stores/v2"]
+
+    def parse(self, response, **kwargs):
+        for store in response.json():
+            store["id"] = store.pop("storeId")
+            address = store["storeAddress"]
+            store["housenumber"] = address.get("streetNo")
+            store["street"] = address.get("street")
+            store["suburb"] = address.get("suburb")
+            store["city"] = address.get("city")
+            province = address.get("province") or {}  # can be JSON null
+            store["province"] = province.get("name")
+            store["country"] = address.get("country")
+            store["postcode"] = address.get("postalCode")
+            item = DictParser.parse(store)
+
+            oh = OpeningHours()
+            for day in store.get("tradingHours", []):
+                # dayId is 1 to 8, 1 being Monday
+                # dayId of 8 is only present on some 24/7 stores so we ignore it
+                if day["dayId"] < 8:
+                    oh.add_range(DAYS[day["dayId"] - 1], day["openTime"], day["closeTime"])
+            item["opening_hours"] = oh
+
+            if store["storeType"] == "CLOTHING":
+                item.update({"brand": "Pick n Pay Clothing", "brand_wikidata": "Q122964352"})
+            elif store["storeType"] == "LIQUOR":
+                item.update({"brand": "Pick n Pay Liquor", "brand_wikidata": "Q122764458"})
+            elif store["storeType"] == "EXPRESS":
+                item.update({"brand": "Pick n Pay Express", "brand_wikidata": "Q122764443"})
+            elif store["storeType"] in ("BOXER", "BOXER PUNCH", "BOXER SUPERSTOR"):
+                item.update({"brand": "Boxer", "brand_wikidata": "Q116586275"})
+            elif store["storeType"] == "BOXER LIQUOR":
+                item.update({"brand": "Boxer Liquors", "brand_wikidata": "Q122766666"})
+            elif store["storeType"] == "BOXER BUILD":
+                item.update({"brand": "Boxer Build", "brand_wikidata": "Q122766671"})
+            # Unhandled: "", "BPFORECOURT", "DAILY", "DC", "FAMILY", "HYPER", "LOCAL", "MARKET",
```

Comment:
```curl https://api.pnp.co.za/stores/v2 | jq '.[].storeType' | sort | uniq -c```

```
      1 ""
      2 "BOXER"
     38 "BOXER BUILD"
      4 "BOXER LIQUOR"
      2 "BOXER PUNCH"
    326 "BOXER SUPERSTOR"
    589 "BPFORECOURT"
    429 "CLOTHING"
      2 "DAILY"
     16 "DC"
    226 "EXPRESS"
    311 "FAMILY"
     21 "HYPER"
     12 "LIQUOR"
     65 "LOCAL"
     47 "MARKET"
     10 "MINI"
      1 "NO_FORMAT"
      3 "ONLINE"
     43 "REGIONAL OFFICE"
    335 "SUPER"
      7 "WHOLESALE"
```

https://www.picknpayinvestor.co.za/store-footprint-format.php mentions a separate "Pick n Pay Market" sub-brand with 36 locations (seemingly now increased to 47 per query results above). Separate Wikidata item and sub-brand needed?

SUPER or FAMILY are each 300+, and we're short on the Pick n Pay supermarket brand by about that much because https://www.picknpayinvestor.co.za/store-footprint-format.php reports about 600 Pick n Pay branded stores across Africa. The Pick n Pay Liquor brand is stated to have over 500 stores per https://www.picknpayinvestor.co.za/store-footprint-format.php but we're only capturing 12, so maybe SUPER or FAMILY are Pick N Pay liquor shops?

CLOTHING here is 429 stores versus 250-odd listed at https://www.picknpayinvestor.co.za/store-footprint-format.php so there appears to be more clothing stores than we want returned?

Reply:
Legally Liquor shops in South Africa have to be a separate entity. Yes, most are co-located (unlike Clothing).